cli_tools/import_precheck/check_disks_linux.go

```diff
@@ -60,47 +87,114 @@ func (c *disksCheck) getName() string {
+func (m *mountPoints) listPhysicalDevicesForMount(mount string) (devices []string) {
+	set := map[string]struct{}{}
+	for _, mp := range m.mounts {
+		if mp.dir == mount {
+			set[mp.hierarchy[0]] = struct{}{}
+		}
 	}
+	for k, _ := range set {
+		devices = append(devices, k)
+	}
+	// sorted for stability in testing.
+	sort.Strings(devices)
+	return devices
+}
 
-	l := lsblkParse(out)
+// listMountPoints returns all mounts.
+func (m *mountPoints) listMountPoints() (mounts []string) {
+	set := map[string]struct{}{}
+	for _, mount := range m.mounts {
+		set[mount.dir] = struct{}{}
+	}
+	for k, _ := range set {
+		mounts = append(mounts, k)
+	}
+	// sorted for stability in testing.
+	sort.Strings(mounts)
+	return mounts
+}
 
-	r.Info(fmt.Sprintf("`lsblk -i` results:\n%s", l.raw))
-	if len(l.devRows) != 1 {
-		r.Warn("translation only supports single block device. Will attempt to determine if translation is possible...")
+// addAll populates the mountPoints data structure with the response from lsblk.
+func (m *mountPoints) addAll(elements []DiskElement, basePath []string) {
+	for _, element := range elements {
+		path := make([]string, len(basePath))
+		copy(path, basePath)
+		path = append(path, element.Name)
+		if element.Mountpoint != "" {
+			m.mounts = append(m.mounts, struct {
+				dir       string
+				hierarchy []string
+			}{element.Mountpoint, path})
+		}
+		if len(element.Children) > 0 {
+			m.addAll(element.Children, path)
+		}
 	}
+}
+
+func (c *disksCheck) run() (r *report, err error) {
+	r = &report{name: c.getName()}
 
-	if l.rootDev == "" {
+	allMounts, err := c.readMounts()
+	if err != nil {
+		r.Warn(fmt.Sprintf("Failed to execute lsblk: %s", err))
+		return r, nil
+	}
+
+	rootDevices := allMounts.listPhysicalDevicesForMount("/")
+	switch len(rootDevices) {
+	case 0:
 		r.Fatal("root filesystem partition not found on any block devices.")
 		return r, nil
+	case 1:
+		r.Info(fmt.Sprintf("root filesystem found on device: %s", rootDevices[0]))
+	default:
+		format := "root filesystem spans multiple physical devices (%s). Translation only supports single block device."
+		r.Fatal(fmt.Sprintf(format, rootDevices))
+		return r, nil
 	}
-	r.Info(fmt.Sprintf("root filesystem on device %q partition %q", l.rootDev, l.rootPart))
 
-	for dev, rows := range l.devRows {
-		for _, row := range rows {
-			if row["MOUNTPOINT"] != "" && dev != l.rootDev {
-				format := "partition %s (%s) is not on the root device %s and will be OMITTED from translation."
-				r.Warn(fmt.Sprintf(format, row["NAME"], row["MOUNTPOINT"], l.rootDev))
+	rootDevice := rootDevices[0]
+	for _, mount := range allMounts.listMountPoints() {
+		if mount == "/" {
+			continue
+		}
+		devices := allMounts.listPhysicalDevicesForMount(mount)
+		switch len(devices) {
+		case 0:
+			// This implies a bug in mountPoints.addAll.
+			panic(fmt.Sprintf("Invalid parse of mount %s", mount))
+		case 1:
+			if devices[0] != rootDevice {
```

Comment:
I feel like it should hard-fail for some specific mount points like /var or /usr (/usr is unlikely these days, but I think some people mount /var on a separate partition/volume).

Reply:
I'm worried about failing the check when there are circumstances where the import *could* have worked.

For example when the image is imported without the guest environment. In that case, there'd be no interactions with the package manager, and therefore probably no writing to /var/log.

(During translation, the disk is mounted as a chroot rather than booting.)